cpp/src/arrow/extension/tensor_internal.h

```diff
@@ -42,4 +47,46 @@ Status IsPermutationValid(const std::vector<int64_t>& permutation) {
   return Status::OK();
 }
 
+ARROW_EXPORT
+inline Status ComputeStrides(const std::shared_ptr<DataType>& value_type,
+                             const std::vector<int64_t>& shape,
+                             const std::vector<int64_t>& permutation,
+                             std::vector<int64_t>* strides) {
+  const auto fixed_width_type =
+      internal::checked_pointer_cast<FixedWidthType>(value_type);
+  if (permutation.empty()) {
+    return internal::ComputeRowMajorStrides(*fixed_width_type.get(), shape, strides);
+  }
+  const int byte_width = value_type->byte_width();
+
+  int64_t remaining = 0;
+  if (!shape.empty() && shape.front() > 0) {
+    remaining = byte_width;
+    for (auto i : permutation) {
+      if (i > 0) {
+        if (internal::MultiplyWithOverflow(remaining, shape[i], &remaining)) {
+          return Status::Invalid(
+              "Strides computed from shape would not fit in 64-bit integer");
+        }
+      }
+    }
+  }
+
+  if (remaining == 0) {
+    strides->assign(shape.size(), byte_width);
+    return Status::OK();
+  }
+
+  strides->push_back(remaining);
+  for (auto i : permutation) {
+    if (i > 0) {
+      remaining /= shape[i];
```

Comment:
Good point. I'll make an issu.

Reply:
How about this: https://github.com/apache/arrow/pull/38008/commits/8e6e99093f493bbe517cdec102d2dca01d2fcac5
Somewhat disturbi

Reply:
Spec states:

  * **permutation**  = indices of the desired ordering of the
    original dimensions, defined as an array.

    The indices contain a permutation of the values [0, 1, .., N-1] where
    N is the number of dimensions. The permutation indicates which
    dimension of the logical layout corresponds to which dimension of the
    physical tensor (the i-th dimension of the logical view corresponds
    to the dimension with number ``permutations[i]`` of the physical tensor).

    Permutation can be useful in case the logical order of
    the tensor is a permutation of the physical order (row-major).

    When logical and physical layout are equal, the permutation will always
    be ([0, 1, .., N-1]) and can therefore be left out.

I understand this as that the logical layout will have to have different strides than the physical layout not just a permutation of physical strides.

Reply:
False alarm. Most `FixedShapeTensor` tests don't need to be refactored. Except two, which is still alarming and we should verify wheter this is introducing a new bug or fixing an old one ot both.
On the positivne note `ComputeStrides` is now easier to reason about.